# Multiple Linear Regression

"D:\Dipti\DA_DS\Python\Prac\MultipleLinearRegression.png"

In [3]:
from warnings import filterwarnings
filterwarnings('ignore')

# Read the 50 Startup dataset

In [4]:
import os
os.chdir('D:/Dipti/DA_DS/Python/Dataframe/')

In [5]:
import pandas as pd
df= pd.read_csv('50_Startups.csv')
df.head()

RND      ADMIN        MKT       STATE     PROFIT
0  165349.20  136897.80  471784.10    New York  192261.83
1  162597.70  151377.59  443898.53  California  191792.06
2  153441.51  101145.55  407934.54     Florida  191050.39
3  144372.41  118671.85  383199.62    New York  182901.99
4  142107.34   91391.77  366168.42     Florida  166187.94

In [6]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50 entries, 0 to 49
Data columns (total 5 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   RND     50 non-null     float64
 1   ADMIN   50 non-null     float64
 2   MKT     50 non-null     float64
 3   STATE   50 non-null     object 
 4   PROFIT  50 non-null     float64
dtypes: float64(4), object(1)
memory usage: 2.1+ KB


In [7]:
df.isna().sum()

RND       0
ADMIN     0
MKT       0
STATE     0
PROFIT    0
dtype: int64

# Seperate X and Y PROFIT(Y) ~ RND, ADMIN, MKT (X)

In [8]:
X = df.drop(columns=['STATE', 'PROFIT'])
X.head()

RND      ADMIN        MKT
0  165349.20  136897.80  471784.10
1  162597.70  151377.59  443898.53
2  153441.51  101145.55  407934.54
3  144372.41  118671.85  383199.62
4  142107.34   91391.77  366168.42

In [9]:
Y = df[['PROFIT']]
Y.head()

PROFIT
0  192261.83
1  191792.06
2  191050.39
3  182901.99
4  166187.94

# Create a Linear Regression Model

In [10]:
from sklearn.linear_model import LinearRegression
model = LinearRegression()
model.fit(X,Y)

LinearRegression()

# Check model intercept and coeff

In [11]:
model.intercept_

array([50122.19298987])

In [12]:
model.coef_

array([[ 0.80571505, -0.02681597,  0.02722806]])

In [13]:
X.columns

Index(['RND', 'ADMIN', 'MKT'], dtype='object')

PROFIT_pred = 50122.19 + 0.8057xRND -0.0268xADMIN + 0.0272xMKT

1. RND increases by 1000 , PROFIT will increase by 805.7
2. ADMIN increases byt 1000, PROFIT will decrease by 26.8
3. MKT increases by 1000, PROFIT will increase by 27.2

# R2 score of model

In [14]:
model.score(X,Y)

0.9507459940683246

# Predict the values X

In [15]:
ypred = model.predict(X)
ypred[0:5]

array([[192521.25289008],
       [189156.76823227],
       [182147.2790962 ],
       [173696.70002553],
       [172139.51418327]])

In [16]:
Y.head()

PROFIT
0  192261.83
1  191792.06
2  191050.39
3  182901.99
4  166187.94

# Evaluate based on different metrics

1. MSE
2. RMSE
3. MAE
4. R2 Score

In [17]:
from sklearn.metrics import mean_squared_error, mean_absolute_error, r2_score

In [20]:
mse = mean_squared_error(Y, ypred)
mse

78417126.0191308

In [21]:
rmse = mse**(1/2)
rmse

8855.344489015139

In [22]:
mae = mean_absolute_error(Y, ypred)
mae

6471.4503961048085

In [23]:
r2 = r2_score(Y, ypred)
r2

0.9507459940683246

# Predicting new values

In [24]:
X.columns

Index(['RND', 'ADMIN', 'MKT'], dtype='object')

In [25]:
xnew = [[200000, 150000, 100000]]
pred = model.predict(xnew)
pred

array([[209965.61419388]])